File: web-client/src/utils/observableUtils.js
// import { defer } from 'rxjs';
// import { flatMap, scan, startWith, retryWhen, delay } from 'rxjs/operators';

// export const retryDelay = (retries, timeDelay) =>
//   retryWhen(errors =>
//     errors.pipe(
//       delay(timeDelay),
//       startWith(1),
//       scan((count, currentErr) => {
//         console.warn('Retry #', count);
//         // console.log(`${count} >= ${retries}`, count >= retries);
//         if (count >= retries) {
//           // console.warn('NO MORE TRIES');
//           throw currentErr;
//         } else {
//           return (count += 1);
//         }
//       })
//     )
//   );

// export const retryableAsync = (funcAsync, retries = 3, timeDelay = 500) =>
//   defer(funcAsync)
//     .pipe(
//       flatMap(async x => await x),
//       retryDelay(retries, timeDelay)
//     )
//     .toPromise();

export const retryableAsync = async (
  asyncFunc,
  retries = 3,
  timeDelay = 500
) => {
  let retryCount = 0;

  while (retryCount < retries) {
    console.log(`Attempt #${retryCount}`);
    try {
      return await asyncFunc();
    } catch (error) {
      console.log(`Failed attempt #${retryCount}`);
      await Promise.delay(timeDelay);
      retryCount++;
    }
  }
};

File: mobile_client/lib/utils/observable_utils.dart
import 'dart:async';

// typedef AsyncFunc<T> = Future<T> Function();
typedef Future<T> AsyncFunc<T>();

Future<T> retryableAsync<T>(
  AsyncFunc<T> asyncFunc, {
  int retries = 3,
  int timeDelay = 500,
}) async {
  int retryCount = 0;

  while (retryCount < retries) {
    print('Attempt #$retryCount');
    try {
      return await asyncFunc();
    } catch (e) {
      print('Failed attempt #$retryCount: $e');
      await Future.delayed(Duration(milliseconds: timeDelay));
      retryCount++;
    }
  }

  return Future.error('Async failed to execute after $retryCount retries');
}
